README.md
#fizzbuzz-rest

Fizzbuzz-rest is a simple fizz-buzz REST server. This server exposes a REST API endpoint that: Accepts five parameters : three integers int1, int2 and limit, and two strings str1 and str2. Returns a list of strings with numbers from 1 to limit, where: all multiples of int1 are replaced by str1, all multiples of int2 are replaced by str2, all multiples of int1 and int2 are replaced by str1str2.

## Installation

This module build a docker image with fizzbuzz-rest server. This application need a redis to store historic. The docker-compose is used to launch fizzbuzz-rest and redis images.

You must install this module in your GOPATH.
``` shell
> go get -u github.com/melkor/fizzbuzz-rest

```

``` shell
> cd ${GOPATH}/src/github.com/melkor/fizzbuzz-rest/
> sudo docker build --tag=fizzbuzz-rest .
> sudo docker-compose up
```

## Exemples

To get the **fizzbuzz** of 1 and 7 with numbers from 1 to 25.
``` shell
> curl 'localhost:8000/fizzbuzz?int1=3&int2=7&limit=25&str1=fizz&str2=fuzz'
["1","2","fizz","4","5","fizz","fuzz","8","fizz","10","11","fizz","13","fuzz","fizz","16","17","fizz","19","20","fizzfuzz","22","23","fizz","25"]
>
```

To get the most frequent request.
``` shell
> curl localhost:8000/mostFrequentRequest
"int1=3\u0026int2=7\u0026limit=25\u0026str1=fizz\u0026str2=fuzz"
>
```

Note: see https://golang.org/pkg/encoding/json/#Marshal for an explanation of **\u0026** instead of **&**



<!--
  506  docker run -p 8000:8000 fizzbuzz-rest
  507  docker build --tag=fizzbuzz-rest .
  508  docker build --tag=fizzbuzz-rest .
  509  docker run -p 8000:8000 fizzbuzz-rest
  510  docker build --tag=fizzbuzz-rest .
  511  docker run -p 8000:8000 fizzbuzz-rest
  512  docker build --tag=fizzbuzz-rest .
  513  docker run -p 8000:8000 fizzbuzz-rest
  514  docker build --tag=fizzbuzz-rest .
  515  docker run -p 8000:8000 fizzbuzz-rest
  516  docker-compose up
  517  curl -L "https://github.com/docker/compose/releases/download/1.23.2/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
  518  chmod +x /usr/local/bin/docker-compose
  519  docker-compose --version
  520  docker-compose up
  521  docker-compose down
-->


debian/changelog
fizzbuzz-rest (0.0~git20190328.dddaa3c-1) unstable; urgency=medium

  * Initial release

 -- Jerome Bourgeois <melkor666free@gmail.com>  Thu, 28 Mar 2019 14:04:10 +0100


app/app_test.go
package app

import (
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/melkor/fizzbuzz-rest/hit"
)

func TestGetFizzBuzz(t *testing.T) {
	testCase := []struct {
		label          string
		int1           string
		int2           string
		limit          string
		str1           string
		str2           string
		expectedStatus int
		expectedBody   string
	}{
		{
			label:          "wrong parameter: int1 is empty",
			expectedStatus: http.StatusBadRequest,
			expectedBody:   "{\"error\":\"int1 must be an integer\"}\n",
		},
		{
			label:          "wrong parameter: int1 not a valid integer",
			int1:           "fizzbuzz",
			expectedStatus: http.StatusBadRequest,
			expectedBody:   "{\"error\":\"int1 must be an integer\"}\n",
		},
		{
			label:          "wrong parameter: int1 == 0",
			int1:           "0",
			int2:           "2",
			limit:          "3",
			expectedStatus: http.StatusBadRequest,
			expectedBody:   "{\"error\":\"int1 can't be lower than 1\"}\n",
		},
		{
			label:          "wrong parameter: int1 == 0",
			int1:           "2",
			int2:           "3",
			limit:          "10",
			str1:           "fizz",
			str2:           "buzz",
			expectedStatus: http.StatusOK,
			expectedBody:   "[\"1\",\"fizz\",\"buzz\",\"fizz\",\"5\",\"fizzbuzz\",\"7\",\"fizz\",\"buzz\",\"fizz\"]\n",
		},
	}
	for _, tc := range testCase {
		req, err := http.NewRequest("GET", "/fizzbuzz", nil)
		if err != nil {
			t.Fatal(err)
		}
		q := req.URL.Query()

		q.Add("int1", tc.int1)
		q.Add("int2", tc.int2)
		q.Add("limit", tc.limit)
		q.Add("str1", tc.str1)
		q.Add("str2", tc.str2)

		req.URL.RawQuery = q.Encode()

		// We create a ResponseRecorder (which satisfies http.ResponseWriter) to record the response.
		rr := httptest.NewRecorder()

		app := New("", "", "")
		app.initializeRoutes()
		app.hit = hit.NewMock()

		// Our handlers satisfy http.Handler, so we can call their ServeHTTP method
		// directly and pass in our Request and ResponseRecorder.
		app.router.ServeHTTP(rr, req)

		// Check the status code is what we expect.
		if status := rr.Code; status != tc.expectedStatus {
			t.Errorf("handler returned wrong status code: got %v want %v",
				status, tc.expectedStatus)
		}

		// Check the response body is what we expect.
		if rr.Body.String() != tc.expectedBody {
			t.Errorf("handler returned unexpected body: got %v want %v",
				rr.Body.String(), tc.expectedBody)
		}

	}
}


hit/hit_test.go
package hit

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestMakeKey(t *testing.T) {
	testCase := []struct {
		label    string
		int1     int
		int2     int
		limit    int
		str1     string
		str2     string
		expected string
	}{
		{
			label:    "test 1",
			int1:     3,
			int2:     5,
			str1:     "fizz",
			str2:     "buzz",
			limit:    15,
			expected: "int1=3&int2=5&limit=15&str1=fizz&str2=buzz",
		},
	}

	for _, tc := range testCase {
		generatedKeyName := makeKey(tc.int1, tc.int2, tc.limit, tc.str1, tc.str2)
		assert.Equal(t, tc.expected, generatedKeyName, tc.label)
	}
}

func TestAdd(t *testing.T) {

	h := New()

	int1 := 3
	int2 := 5
	limit := 20
	str1 := "fizz"
	str2 := "fuzz"

	expectedMostFrequentedRequest := makeKey(int1, int2, limit, str1, str2)

	for i := 0; i < 2; i++ {
		h.Add(int1, int2, limit, str1, str2)

	}
	h.Add(1, 2, 3, "puzzle", "buble")

	assert.Equal(t, expectedMostFrequentedRequest, h.GetMostFrequentRequest(), "")
}


hit/redis_test.go
package hit

import (
	"testing"

	"github.com/alicebob/miniredis"
	"github.com/go-redis/redis"
	"github.com/stretchr/testify/assert"
)

func TestCache(t *testing.T) {
	mr, err := miniredis.Run()
	if err != nil {
		panic(err)
	}

	c := &Cache{
		client: redis.NewClient(&redis.Options{
			Addr: mr.Addr(),
		}),
	}

	int1 := 3
	int2 := 5
	limit := 20
	str1 := "fizz"
	str2 := "fuzz"

	expectedMostFrequentedRequest := makeKey(int1, int2, limit, str1, str2)

	for i := 0; i < 2; i++ {
		_, err := c.Add(int1, int2, limit, str1, str2)
		assert.NoError(t, err)
	}
	_, err = c.Add(1, 2, 3, "puzzle", "buble")
	assert.NoError(t, err)

	val, err := c.GetMostFrequentRequest()
	if err != nil {
		panic(err)
	}
	assert.Equal(t, expectedMostFrequentedRequest, val, "")

}


hit/redis.go
package hit

import "github.com/go-redis/redis"

// We use a redis to store "hits" in a sorted set:
// - we can use more than one fizzbuzz-rest if we need load balancing to
//   improve performance
// - we don't need to check concurrent access because of redis provides atomic
//   functions
// - we can make this historic persistant if fizzbuzz-rest crash

//ZSetKey is the key name of ordered set
const ZSetKey = "fizzbuzz-hit"

//Cache is used to store hit into a Redis
type Cache struct {
	client *redis.Client
}

//NewCache initialize a Cache hti object
func NewCache(addr, password string, db int) *Cache {

	c := &Cache{
		client: redis.NewClient(&redis.Options{
			Addr:     addr,
			Password: password,
			DB:       db,
		}),
	}
	return c
}

//Add is used to add a request in Hit
func (c *Cache) Add(int1, int2, limit int, str1, str2 string) (float64, error) {

	// add a member in the ZSET or update the score of this member
	member := makeKey(int1, int2, limit, str1, str2)
	return c.client.ZIncr(
		ZSetKey,
		redis.Z{
			Score:  1,
			Member: member,
		},
	).Result()
}

//GetMostFrequentRequest return the most frequebt request
func (c *Cache) GetMostFrequentRequest() (string, error) {

	// to get the hightest score in the ZSet
	//  ZRANGEBYSCORE fizzbuzz-hit +inf -inf 0 1
	vals, err := c.client.ZRevRangeByScore(
		ZSetKey,
		redis.ZRangeBy{
			Max:    "+inf",
			Min:    "-inf",
			Offset: 0,
			Count:  1,
		},
	).Result()

	if err != nil {
		return "", err
	}

	if len(vals) > 0 {
		return vals[0], nil
	}

	return "", nil
}


hit/hit_mock.go
package hit

//This package is a mock of Cache (see redis.go) used in units test of the
//application (app_test.go).

//Hit represents an object store hit in memory
type Hit struct {
	higherHit           float64            //the score of the mostFrequentRequest
	mostFrequentRequest string             //the most frequent request
	requests            map[string]float64 //historic of all requests
}

//NewMock initialize a Hit mocker object
func NewMock() *Hit {
	h := &Hit{}
	h.requests = make(map[string]float64)
	return h
}

//Add is used to add a request in Hit
func (h *Hit) Add(int1, int2, limit int, str1, str2 string) (float64, error) {
	key := makeKey(int1, int2, limit, str1, str2)

	// if request is not present in historic, we initialize the score of it to 0
	// else we add 1 to that score
	if _, exists := h.requests[key]; exists {
		h.requests[key]++
	} else {
		h.requests[key] = 1
	}

	// if score it higer thant higherHit, this request become the most
	// frequent request
	if h.requests[key] > h.higherHit {
		h.higherHit = h.requests[key]
		h.mostFrequentRequest = key
	}

	return h.requests[key], nil
}

//GetMostFrequentRequest return the most frequebt request
func (h *Hit) GetMostFrequentRequest() (string, error) {
	return h.mostFrequentRequest, nil
}


fizzbuzz/fizzbuzz.go
package fizzbuzz

import (
	"errors"
	"strconv"
)

//Fizzbuzz must return a list of strings with numbers from 1 to limit, where:
// all multiples of int1 are replaced by str1, all multiples of int2 are
// replaced by str2, all multiples of int1 and int2 are replaced by str1str2
func Fizzbuzz(int1, int2, limit int, str1, str2 string) ([]string, error) {

	// check parameters int1 int2 and limit are valides to avoid divid by 0 for
	// example
	if int1 < 1 {
		return []string{}, errors.New("int1 can't be lower than 1")
	}

	if int2 < 1 {
		return []string{}, errors.New("int2 can't be lower than 1")
	}

	if limit < 1 {
		return []string{}, errors.New("limit can't be lower than 1")
	}

	ret := make([]string, 0, limit-1)

	// a simple fizzbuzz algorithm
	n := 1
	for n <= limit {
		if n%int1 == 0 && n%int2 == 0 {
			ret = append(ret, str1+str2)
		} else if n%int1 == 0 {
			ret = append(ret, str1)
		} else if n%int2 == 0 {
			ret = append(ret, str2)
		} else {
			ret = append(ret, strconv.Itoa(n))
		}
		n++
	}

	return ret, nil
}


cmd/fizzbuzz-rest/main.go
package main

// this package check command line paramters and launch application

import (
	"fmt"
	"os"

	"github.com/melkor/fizzbuzz-rest/app"
	log "github.com/sirupsen/logrus"
	"github.com/spf13/pflag"
)

var (
	address       = pflag.String("address", ":8000", "The listening address of the web server")
	cacheAddress  = pflag.String("cache-address", "localhost:6379", "Address of the redis server used to store hits")
	cachePassword = pflag.String("cache-password", "", "Password of the redis server")
	debug         = pflag.BoolP("debug", "d", false, "Activate debug mode")
	help          = pflag.BoolP("help", "h", false, "Display usage message")
	verbose       = pflag.BoolP("verbose", "v", false, "Activate verbose mode")
)

// display help message and exit
func displayHelp() {
	fmt.Println(os.Args[0])
	fmt.Println("Web server that will expose a REST API to execute fizz-buzz.")
	pflag.PrintDefaults()
}

func main() {
	pflag.Parse()

	if *help {
		displayHelp()
		os.Exit(0)
	}

	if *debug {
		log.SetLevel(log.DebugLevel)
	} else if *verbose {
		log.SetLevel(log.InfoLevel)
	}

	app := app.New(*address, *cacheAddress, *cachePassword)
	app.Run()
}


app/app.go
package app

// this package handle endpoints of the application

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/gorilla/mux"
	"github.com/melkor/fizzbuzz-rest/fizzbuzz"
	"github.com/melkor/fizzbuzz-rest/hit"
	log "github.com/sirupsen/logrus"
)

//App struct used to handle application
type App struct {
	address string
	hit     hit.Handler
	router  *mux.Router
}

//New initialize application
func New(listenAddress, cacheAddress, cachePassword string) *App {

	//only check if an address was set, the validity check
	// will be done by http.ListenAndServ
	if listenAddress == "" {
		listenAddress = ":8000"
	}

	if cacheAddress == "" {
		cacheAddress = "localhost:6379"
	}

	a := &App{
		address: listenAddress,
		hit:     hit.NewCache(cacheAddress, cachePassword, 0),
		router:  mux.NewRouter(),
	}

	return a
}

//getFizzBuzz is the endpoint behind the route "/fizzbuzz"
// which will call fizzbuzz func and return result as a json message
func (a *App) getFizzBuzz(w http.ResponseWriter, r *http.Request) {
	log.Debugln("received a fizzbuzz request")

	w.Header().Set("Content-Type", "application/json")

	params := mux.Vars(r)
	paramsIntVal := make(map[string]int)

	// this function is called only if all parameters are set in
	// the querystring, we don't need to check if keys (int1 ,int2,
	// limit, str1 and str2) are present in params map

	//convert into integer the parameters from querystring
	// which must be handled as integer
	for _, name := range []string{"int1", "int2", "limit"} {
		val, err := strconv.Atoi(params[name])
		log.Debugln("	param: ", name, ", value: ", val)
		if err != nil {
			w.WriteHeader(http.StatusBadRequest)
			log.Errorln("query error : ", err)
			err := json.NewEncoder(w).Encode(map[string]string{"error": name + " must be an integer"})
			if err != nil {
				log.Errorln("jsonize error : ", err)
			}
			return
		}
		paramsIntVal[name] = val
	}

	log.Debugln("	param: str1, value: ", params["str1"])
	log.Debugln("	param: str2, value: ", params["str2"])

	// call fizz-buzz function
	result, err := fizzbuzz.Fizzbuzz(
		paramsIntVal["int1"],
		paramsIntVal["int2"],
		paramsIntVal["limit"],
		params["str1"],
		params["str2"],
	)

	if err != nil {
		w.WriteHeader(http.StatusBadRequest)
		log.Errorln("fizzbuzz error : ", err)
		err := json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
		if err != nil {
			log.Errorln("jsonize error : ", err)
		}
		return
	}

	log.Debugln("	result: ", result)

	err = json.NewEncoder(w).Encode(result)
	if err != nil {
		log.Errorln("jsonize error : ", err)
	}
	log.Debugln("Done...")

	log.Debugln("Add request into historic")
	score, err := a.hit.Add(
		paramsIntVal["int1"],
		paramsIntVal["int2"],
		paramsIntVal["limit"],
		params["str1"],
		params["str2"],
	)

	if err != nil {
		log.Errorln("add hit error : ", err)
	}

	log.Debugln("hit score: ", score)
}

func (a *App) getMostFrequentRequest(w http.ResponseWriter, r *http.Request) {
	log.Debugln("received an audience request")

	w.Header().Set("Content-Type", "application/json")

	result, err := a.hit.GetMostFrequentRequest()

	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		log.Errorln("most frequent request error : ", err)
		err := json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
		if err != nil {
			log.Errorln("jsonize error : ", err)
		}
		return
	}

	err = json.NewEncoder(w).Encode(result)
	if err != nil {
		log.Errorln("jsonize error : ", err)
	}

	log.Debugln("	result: ", result)
}

//add http route into this func
func (a *App) initializeRoutes() {
	log.Infoln("Initialize routes")

	//route "GET" to fizzbuzzaddr
	//Note: all parameters must be set to use this route
	log.Debugln("Add route to /fizzbuzz endpoint")
	a.router.HandleFunc("/fizzbuzz", a.getFizzBuzz).
		Methods("GET").
		Queries("int1", "{int1}").
		Queries("int2", "{int2}").
		Queries("limit", "{limit}").
		Queries("str1", "{str1}").
		Queries("str2", "{str2}")

	//endpoint to most frequent request
	log.Debugln("Add route to /mostFrequentRequest endpoint")
	a.router.HandleFunc("/mostFrequentRequest", a.getMostFrequentRequest)

}

//Run function launch application
func (a *App) Run() {
	log.Infoln("Launch application")
	a.initializeRoutes()
	log.Fatal(http.ListenAndServe(a.address, a.router))
}
